Change adoptNode() for a DocumentFragment with host

rniwa opened the pull request:
#### ab3c0b8cf2a411231da9526614ad5167971dcdd5
<pre>
Change adoptNode() for a DocumentFragment with host
<a href="https://bugs.webkit.org/show_bug.cgi?id=204980">https://bugs.webkit.org/show_bug.cgi?id=204980</a>

Reviewed by Darin Adler.

Disallow adoptNode on a TemplateContentDocumentFragment to align with the spec.

Also re-introduce a fast path in containsIncludingHostElements that was removed
in <a href="https://commits.webkit.org/r272703">https://commits.webkit.org/r272703</a> to handle this particular case.

* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/adoption.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-template-element/template-element/template-content-hierarcy-expected.txt:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::containsIncludingHostElements):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::adoptNode): Return early when a node is TemplateContentDocumentFragment (i.e.
it&apos;s not a shadow root but has a host).
* Source/WebCore/dom/TemplateContentDocumentFragment.h:
(isType): Added.

Canonical link: <a href="https://commits.webkit.org/252098@main">https://commits.webkit.org/252098@main</a>
</pre>

darinadler commented on Source/WebCore/dom/TemplateContentDocumentFragment.h:
```diff
@@ -55,3 +55,8 @@ class TemplateContentDocumentFragment final : public DocumentFragment {
 } // namespace WebCore
+
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::TemplateContentDocumentFragment)
+    static bool isType(const WebCore::Node& node) { return node.isDocumentFragment() && static_cast<const WebCore::DocumentFragment&>(node).isTemplateContent(); }
```
Can this use `dynamicDowncast` and call the other `isType` overload instead of `static_cast` and repeating what it does?

rniwa replied:
Sure, done.

rniwa committed:
Change adoptNode() for a DocumentFragment with host
https://bugs.webkit.org/show_bug.cgi?id=204980

Reviewed by Darin Adler.

Disallow adoptNode on a TemplateContentDocumentFragment to align with the spec.

Also re-introduce a fast path in containsIncludingHostElements that was removed
in https://commits.webkit.org/r272703 to handle this particular case.

* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/adoption.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-template-element/template-element/template-content-hierarcy-expected.txt:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::containsIncludingHostElements):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::adoptNode): Return early when a node is TemplateContentDocumentFragment (i.e.
it's not a shadow root but has a host).
* Source/WebCore/dom/TemplateContentDocumentFragment.h:
(isType): Added.

Canonical link: https://commits.webkit.org/252098@main
